### ISSUE TITLE
Grammar

### ISSUE THREAD

**Issue by Jeffreyfrmn** [CG020/team_lightning#3]
Get our grammar down!

**Comment by CG020** [comment]
Add the grammar as document in repo

**Comment by CG020** [comment]
Grammar Draft 1

<statement> ::= <variable> | <funCall> | <function> | <conditional> | <loop>
<expr> ::= <int> | <real> | <char> | <string> | <bool> | <variable> | <expr> <operator> <expr>
| <funCall>
<variable> ::= <name> !  | <name> = <expr> !
<name>::= all valid names regex pattern (no quotes)

<function> ::= fun <name> (<params>) = <statement>
<funCall> ::= <name> (<args>) !
<params> ::= <variable> | <variable> , <params>
<args> ::= <expr> | <expr> , <args>

<conditional> ::= if <condition> ? <statement> ! | if <condition> ? <statement> ! else
 <statement> !
<condition> ::= <expr> | nono <expr>

<loop> ::= while <condition> ? <statement> !

<operator> ::= add | sub | mult | div | mod
<compare> ::= also | ore | less | more | more ore same | less ore same | same

<int> ::= includes all numbers no ‘.’ included
<real> ::=  [+-]?([0-9]*[.])?[0-9]+
<char> ::= “[:alphanum:]{0:1}” 
<string> ::= “ all strings ”
<bool> ::= kachow | kachiga

**Comment by CG020** [comment]
\<statement> ::= \<variable> | \<funCall> | \<function> | \<conditional> | \<loop>
\<expr> ::= \<int> | \<real> | \<char> | \<string> | \<bool> | \<variable> | \<expr> \<operator> \<expr> | \<funCall>
\<variable> ::= \<name> !  | \<name> = \<expr> !
\<name>::= all valid names regex pattern (no quotes)

\<function> ::= fun \<name> (\<params>) = \<statement>
\<funCall> ::= \<name> (\<args>) !
\<params> ::= \<variable> | \<variable> , \<params>
\<args> ::= \<expr> | \<expr> , \<args>

\<conditional> ::= if \<condition> ? \<statement> ! | if \<condition> ? \<statement> ! else  \<statement> !
\<condition> ::= \<expr> | nono \<expr>

\<loop> ::= while \<condition> ? \<statement> !

\<operator> ::= add | sub | mult | div | mod
\<compare> ::= also | ore | less | more | more ore same | less ore same | same

\<int> ::= includes all numbers no ‘.’ included
\<real> ::=  [+-]?([0-9]*[.])?[0-9]+
\<char> ::= “[:alphanum:]{0:1}” 
\<string> ::= “ all strings ”
\<bool> ::= kachow | kachiga

**Comment by Jeffreyfrmn** [comment]
@CG020 Hey I noticed we don't have any data structures in our grammar (list, tuple, map).

**Comment by CG020** [comment]
You're right i updated with all three - the delimiters i just kept as standard we can change later if we want 
[TL_grammar.txt](https://github.com/CG020/team_lightning/files/14899443/TL_grammar.txt)

**Comment by Jeffreyfrmn** [comment]
Changing grammer character from string of length 1 to <character> and all escaped characters = <\escape_character> like newline -> <\n> or null -> <\0>.

The regex logic is getting too complex for my tastes.

**Comment by Jeffreyfrmn** [comment]
With some quick statement tests the variable initialization sucks to regex as well. We can round back to it but temporarily lets initialize variables that will have an initial null value to:

variable_name = !
